From 09dd0092618b94f538383308e9e505105d05cfaa Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Mon, 27 May 2013 12:58:33 +0700
Subject: Dalvik: peelLoop uses operator of map incorrectly

BZ: 110748

STL map [] operator always creates a new entry if used key is not
in the map while peelLoop expects that [] returns 0 in this case.
Patch uses count() member function to check whether the key is in map.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ie7d78396a79c3b44643e93a59fdb5f70ad9883d4
Orig-MCG-Change-Id: I31467fe8832ba42038a16d24479508e6f35fe036
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/BBOptimization.cpp |   12 ++++--------
 1 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/vm/compiler/BBOptimization.cpp b/vm/compiler/BBOptimization.cpp
index 29e0935..09f2342 100644
--- a/vm/compiler/BBOptimization.cpp
+++ b/vm/compiler/BBOptimization.cpp
@@ -1655,13 +1655,11 @@ void peelLoop (CompilationUnit *cUnit, LoopInformation *info)
         //Transform the taken and fallthrough using the associations
         if (taken != 0)
         {
-            BasicBlock *newTaken = associations[taken];
-
             //Update the association if we have one
-            if (newTaken != 0)
+            if (associations.count(taken) > 0)
             {
                 //Now update the link
-                copy->taken = newTaken;
+                copy->taken = associations[taken];
             }
             else
             {
@@ -1679,13 +1677,11 @@ void peelLoop (CompilationUnit *cUnit, LoopInformation *info)
 
         if (fallThrough != 0)
         {
-            BasicBlock *newFallThrough = associations[fallThrough];
-
             //Update the association if we have one
-            if (newFallThrough != 0)
+            if (associations.count(fallThrough) > 0)
             {
                 //Now update the link
-                copy->fallThrough = newFallThrough;
+                copy->fallThrough = associations[fallThrough];
             }
             else
             {
-- 
1.7.4.1

